File: go/mechanisms/evm/batch-settlement/facilitator/claim.go
package facilitator

import (
	"context"
	"fmt"
	"math/big"
	"strings"

	x402 "github.com/x402-foundation/x402/go"
	"github.com/x402-foundation/x402/go/mechanisms/evm"
	batchsettlement "github.com/x402-foundation/x402/go/mechanisms/evm/batch-settlement"
	"github.com/x402-foundation/x402/go/types"
)

// ExecuteClaimWithSignature executes a batch claim with receiverAuthorizer signature.
// If ClaimAuthorizerSignature is absent from the payload, the authorizerSigner
// auto-signs the ClaimBatch digest.
func ExecuteClaimWithSignature(
	ctx context.Context,
	signer evm.FacilitatorEvmSigner,
	payload *batchsettlement.BatchSettlementClaimPayload,
	requirements types.PaymentRequirements,
	authorizerSigner batchsettlement.AuthorizerSigner,
) (*x402.SettleResponse, error) {
	network := x402.Network(requirements.Network)

	if len(payload.Claims) == 0 {
		return nil, x402.NewSettleError(ErrInvalidClaimPayload, "", network, "",
			"no claims provided")
	}

	// Resolve signature — auto-sign if absent
	var sigBytes []byte
	if payload.ClaimAuthorizerSignature != "" {
		var err error
		sigBytes, err = evm.HexToBytes(payload.ClaimAuthorizerSignature)
		if err != nil {
			return nil, x402.NewSettleError(ErrInvalidClaimPayload, "", network, "",
				fmt.Sprintf("invalid claim authorizer signature: %s", err))
		}
	} else {
		// Verify authorizer address matches all claims' receiverAuthorizer
		for _, claim := range payload.Claims {
			if !strings.EqualFold(claim.Voucher.Channel.ReceiverAuthorizer, authorizerSigner.Address()) {
				return nil, x402.NewSettleError(ErrAuthorizerAddressMismatch, "", network, "",
					fmt.Sprintf("claim receiverAuthorizer %s does not match authorizerSigner %s",
						claim.Voucher.Channel.ReceiverAuthorizer, authorizerSigner.Address()))
			}
		}
		// Auto-sign
		var err error
		sigBytes, err = authorizerSigner.SignClaimBatch(ctx, payload.Claims, string(network))
		if err != nil {
			return nil, x402.NewSettleError(ErrClaimTransactionFailed, "", network, "",
				fmt.Sprintf("failed to sign claim batch: %s", err))
		}
	}

	claimArgs := buildVoucherClaimArgs(payload.Claims)

	// Simulate the transaction before submitting
	if _, simErr := signer.ReadContract(
		ctx,
		batchsettlement.BatchSettlementAddress,
		batchsettlement.BatchSettlementClaimWithSignatureABI,
		"claimWithSignature",
		claimArgs,
		sigBytes,
	); simErr != nil {
		return &x402.SettleResponse{ //nolint:nilerr // simulation failure → error encoded in response
			Success:      false,
			ErrorReason:  ErrClaimSimulationFailed,
			ErrorMessage: simErr.Error(),
			Transaction:  "",
			Network:      network,
		}, nil
	}

	txHash, err := signer.WriteContract(
		ctx,
		batchsettlement.BatchSettlementAddress,
		batchsettlement.BatchSettlementClaimWithSignatureABI,
		"claimWithSignature",
		claimArgs,
		sigBytes,
	)
	if err != nil {
		return nil, x402.NewSettleError(ErrClaimTransactionFailed, "", network, "",
			fmt.Sprintf("claimWithSignature transaction failed: %s", err))
	}

	receipt, err := signer.WaitForTransactionReceipt(ctx, txHash)
	if err != nil {
		return nil, x402.NewSettleError(ErrWaitForReceipt, txHash, network, "",
			fmt.Sprintf("failed waiting for claimWithSignature receipt: %s", err))
	}
	if receipt.Status != evm.TxStatusSuccess {
		return nil, x402.NewSettleError(ErrTransactionReverted, txHash, network, "",
			"claimWithSignature transaction reverted")
	}

	return &x402.SettleResponse{
		Success:     true,
		Transaction: txHash,
		Network:     network,
	}, nil
}

// buildVoucherClaimArgs builds the Solidity-compatible VoucherClaim[] argument for claim calls.
func buildVoucherClaimArgs(claims []batchsettlement.BatchSettlementVoucherClaim) interface{} {
	type VoucherStruct struct {
		Channel            ContractChannelConfigTuple
		MaxClaimableAmount *big.Int
	}
	type VoucherClaimStruct struct {
		Voucher      VoucherStruct
		Signature    []byte
		TotalClaimed *big.Int
	}

	result := make([]VoucherClaimStruct, len(claims))
	for i, claim := range claims {
		maxClaimable, _ := new(big.Int).SetString(claim.Voucher.MaxClaimableAmount, 10)
		totalClaimed, _ := new(big.Int).SetString(claim.TotalClaimed, 10)
		sigBytes, _ := evm.HexToBytes(claim.Signature)

		channelTuple := ToContractChannelConfig(claim.Voucher.Channel)

		result[i] = VoucherClaimStruct{
			Voucher: VoucherStruct{
				Channel:            channelTuple,
				MaxClaimableAmount: maxClaimable,
			},
			Signature:    sigBytes,
			TotalClaimed: totalClaimed,
		}
	}
	return result
}

File: typescript/packages/mechanisms/hedera/src/batch-settlement/facilitator/claim.ts
import type { SettleResponse, PaymentRequirements } from "@x402/core/types";
import { getAddress } from "viem";
import type { FacilitatorHederaBatchSigner } from "../signer";
import type { AuthorizerSigner, BatchSettlementClaimPayload } from "../types";
import { batchSettlementABI } from "../abi";
import { HEDERA_GAS, getBatchSettlementDeployment } from "../constants";
import { signClaimBatch } from "../authorizerSigner";
import * as Errors from "../errors";
import { runContractSettlement, truncate } from "../transport";
import { toContractChannelConfig } from "./utils";

/** Options for claim execution. */
export type ClaimExecutionOptions = {
  simulateBeforeSend?: boolean;
  /** Gas limit override; defaults to `claimBase + claimPerVoucher × n`. */
  gas?: bigint;
};

/**
 * Converts voucher claims into the onchain tuple format expected by `claimWithSignature()`.
 *
 * @param claims - Typed voucher claims with channel config, amounts, and signatures.
 * @returns Contract-ready VoucherClaim argument array.
 */
export function buildVoucherClaimArgs(claims: BatchSettlementClaimPayload["claims"]) {
  return claims.map(c => ({
    voucher: {
      channel: toContractChannelConfig(c.voucher.channel),
      maxClaimableAmount: BigInt(c.voucher.maxClaimableAmount),
    },
    signature: c.signature,
    totalClaimed: BigInt(c.totalClaimed),
  }));
}

/**
 * Default gas for a claim batch of `n` vouchers.
 *
 * @param n - Number of vouchers in the batch.
 * @returns Gas limit.
 */
export function claimGasFor(n: number): bigint {
  return HEDERA_GAS.claimBase + HEDERA_GAS.claimPerVoucher * BigInt(n);
}

/**
 * Submits a batch claim via `claimWithSignature()`.
 *
 * When `claimAuthorizerSignature` is present in the payload it is used directly. When absent the
 * facilitator signs the `ClaimBatch` digest with `authorizerSigner`, after verifying that every
 * claim's `receiverAuthorizer` matches `authorizerSigner.address`.
 *
 * @param signer - Facilitator signer used to submit the claim transaction.
 * @param payload - Claim payload containing voucher claims and optional authorizer signature.
 * @param requirements - Payment requirements for network identification.
 * @param authorizerSigner - Optional dedicated key for producing `ClaimBatch` signatures.
 * @param options - Simulation / gas options.
 * @returns A {@link SettleResponse} with the transaction id on success.
 */
export async function executeClaimWithSignature(
  signer: FacilitatorHederaBatchSigner,
  payload: BatchSettlementClaimPayload,
  requirements: PaymentRequirements,
  authorizerSigner: AuthorizerSigner | undefined,
  options: ClaimExecutionOptions = {},
): Promise<SettleResponse> {
  const network = requirements.network;
  if (!Array.isArray(payload.claims) || payload.claims.length === 0) {
    return { success: false, errorReason: Errors.ErrInvalidPayloadType, transaction: "", network };
  }
  const claimArgs = buildVoucherClaimArgs(payload.claims);
  const contractAddr = getAddress(getBatchSettlementDeployment(network).settlement);

  let sig = payload.claimAuthorizerSignature;

  if (!sig) {
    if (!authorizerSigner) {
      return {
        success: false,
        errorReason: Errors.ErrAuthorizerNotConfigured,
        transaction: "",
        network,
      };
    }
    for (const claim of payload.claims) {
      if (
        getAddress(claim.voucher.channel.receiverAuthorizer) !==
        getAddress(authorizerSigner.address)
      ) {
        return {
          success: false,
          errorReason: Errors.ErrAuthorizerAddressMismatch,
          transaction: "",
          network,
        };
      }
    }
    sig = await signClaimBatch(authorizerSigner, payload.claims, network);
  }

  if (options.simulateBeforeSend ?? true) {
    try {
      await signer.simulateContract({
        address: contractAddr,
        abi: batchSettlementABI,
        functionName: "claimWithSignature",
        args: [claimArgs, sig],
      });
    } catch (e) {
      return {
        success: false,
        errorReason: Errors.ErrClaimSimulationFailed,
        errorMessage: truncate(e instanceof Error ? e.message : String(e)),
        transaction: "",
        network,
      };
    }
  }

  return runContractSettlement(
    () =>
      signer.executeContract({
        address: contractAddr,
        abi: batchSettlementABI,
        functionName: "claimWithSignature",
        args: [claimArgs, sig],
        gas: options.gas ?? claimGasFor(payload.claims.length),
      }),
    network,
    undefined,
    { failedStatusReason: Errors.ErrClaimTransactionFailed },
  );
}
